File: meetings.rb
def merge_ranges(meeting_times)
  start_times = []
  end_times = []
  current_meeting = {}
  merged_times = []
  current_start_time = 0

  meeting_times.each_with_index do |times, index|
    current_meeting = meeting_times[index]
    start_times << current_meeting[:start_time]
    end_times << current_meeting[:end_time]
  end

  start_times.sort!
  end_times.sort!

  current_start_time = start_times.first
  start_times.each_with_index do |time, index|
    index + 1
    if index == end_times.length - 1
      merged_times << { start_time: current_start_time, end_time: end_times[index] }
    elsif end_times[index] < start_times[index + 1]
      merged_times << { start_time: current_start_time, end_time: end_times[index] }
      current_start_time = start_times[index + 1]
    end
  end

  merged_times
end

merge_ranges(
  [
    { start_time: 0,  end_time: 1 },
    { start_time: 3,  end_time: 5 },
    { start_time: 4,  end_time: 8 },
    { start_time: 10, end_time: 12 },
    { start_time: 9,  end_time: 10 }
  ])

File: meetings.js
function mergeRanges(meetingTimes) {
  let startTimes = [];
  let endTimes = [];
  let currentMeeting = {};
  let mergedTimes = [];
  let currentStartTime = 0;

  for (var time in meetingTimes) {
    currentMeeting = meetingTimes[time];
    startTimes.push(currentMeeting['startTime']);
    endTimes.push(currentMeeting['endTime']);
  }

  startTimes.sort(function(a,b) { return a - b; });
  endTimes.sort(function(a,b) { return a - b; });

  currentStartTime = startTimes[0];
  for (var i = 0; i < startTimes.length; i++) {
    if (endTimes[i] < startTimes[i + 1]) {
      mergedTimes.push({startTime: currentStartTime, endTime: endTimes[i]});
      currentStartTime = startTimes[i + 1];
    } else if (i === endTimes.length - 1) {
      mergedTimes.push({startTime: currentStartTime, endTime: endTimes[i]});
    }
  }

  return mergedTimes;
}

mergeRanges(
  [
    {startTime: 0,  endTime: 1},
    {startTime: 3,  endTime: 5},
    {startTime: 4,  endTime: 8},
    {startTime: 10, endTime: 12},
    {startTime: 9,  endTime: 10},
  ])
